cgi_client.go

```diff
@@ -0,0 +1,195 @@
+
+	res, err := c.httpClient.Do(req)
+	defer func() {
+		if res != nil && res.Body != nil {
+			res.Body.Close()
+		}
+	}()
+
+	if err != nil {
+		if strings.Contains(err.Error(), "use of closed network connection") {
+			return fmt.Errorf("timeout occured while accessing %v", loginURL)
+		}
+		return err
+	}
+	return nil
+}
+
+// getBugzillaLoginToken returns Bugzilla_login_token input field value. Requires login cookie to be set
+func (c *bugzillaCGIClient) getBugzillaLoginToken(loginURL string) (loginToken string, err error) {
+	req, err := newHTTPRequest("GET", loginURL, nil)
+	if err != nil {
+		return "", err
+	}
+
+	res, err := c.httpClient.Do(req)
+	defer func() {
+		if res != nil && res.Body != nil {
+			res.Body.Close()
+		}
+	}()
+	if err != nil {
+		if strings.Contains(err.Error(), "use of closed network connection") {
+			return "", fmt.Errorf("timeout occured while accessing %v", loginURL)
+		}
+		return "", err
+	}
+	r := regexp.MustCompile(`name="Bugzilla_login_token"\s+value="(?P<value>[\d\w-]+)"`)
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return "", err
+	}
+	match := r.FindStringSubmatch(string(body))
+	a := make(map[string]string)
+	for i, name := range r.SubexpNames() {
+		a[name] = match[i]
+	}
+	return a["value"], nil
+}
+
+// Login allows to login using Bugzilla CGI API
+func (c *bugzillaCGIClient) login() (err error) {
+	log.Printf("Authenticating to %q ...", c.bugzillaAddr)
+
+	u, err := url.Parse(c.bugzillaAddr)
+	if err != nil {
+		return err
+	}
+	u.Path = "index.cgi"
+	loginURL := u.String()
+
+	err = c.setBugzillaLoginCookie(loginURL)
+	if err != nil {
+		return err
+	}
+
+	loginToken, err := c.getBugzillaLoginToken(loginURL)
+	if err != nil {
+		return err
+	}
+
+	data := url.Values{}
+	data.Set("Bugzilla_login", c.bugzillaLogin)
+	data.Set("Bugzilla_password", c.bugzillaPassword)
+	data.Set("Bugzilla_login_token", loginToken)
+
+	req, err := newHTTPRequest("POST", loginURL, strings.NewReader(data.Encode()))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Accept", "text/html")
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	res, err := c.httpClient.Do(req)
+	defer func() {
+		if res != nil && res.Body != nil {
+			res.Body.Close()
+		}
+	}()
+	if err != nil {
+		if strings.Contains(err.Error(), "use of closed network connection") {
+			return fmt.Errorf("timeout occured while accessing %v", loginURL)
+		}
+		return err
+	}
+
+	return nil
+}
+
+func (c *bugzillaCGIClient) GetCookies() []*http.Cookie {
+	u, err := url.Parse(c.bugzillaAddr)
+	if err != nil {
+		panic(err)
+	}
+	cookies := c.httpClient.Jar.Cookies(u)
+	return cookies
+}
+
+func (c *bugzillaCGIClient) SetCookies(cookies []*http.Cookie) {
+	u, _ := url.Parse(c.bugzillaAddr)
+	c.httpClient.Jar.SetCookies(u, cookies)
+}
+
+func (c *bugzillaCGIClient) authenticated(f func() (*http.Response, error)) (*http.Response, error) {
+	res, err := f()
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
```

Comment:
Do you need to check res != nil?

Reply:
response is `nil` only when `err != nil` (which is handled above)

Reply:
I guess I really meant to ask if you really wanted to copy the sort of idiomatic:
        defer func() {
		if res != nil && res.Body != nil {
			res.Body.Close()
		}
	}()

Reply:
no, I meant if `err == nil`, then there is a zero chance the `res` will be `nil`. Your ask is legit, but not necessary :-) 

I will fix that.